source/common/config/watch_map.cc

```diff
@@ -54,8 +54,17 @@ absl::flat_hash_set<Watch*> WatchMap::watchesInterestedIn(const std::string& res
 void WatchMap::onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources,
                               const std::string& version_info) {
   if (watches_.empty()) {
-    ENVOY_LOG(warn, "WatchMap::onConfigUpdate: there are no watches!");
-    return;
+    if (resources.empty()) {
+      // We have no watches, and the update contained no resources. This can happen when Envoy
+      // unregisters from a resource that's removed from the server as well. For example,
+      // a deleted cluster triggers un-watching the ClusterLoadAssignment watch, and at the
+      // same time the xDS server sends an empty list of ClusterLoadAssignment resources.
+      return;
+    } else {
+      // We have no watches, but the update contained resources. This should not happen.
+      ENVOY_LOG(warn, "Rejecting non-empty update for unwatched type URL");
```

Comment:
I'm curious about this behavior. Let's say the management servers sends on ADS some `ClusterLoadAssignment` for an unknown cluster. I think the behavior we want is to ignore rather than reject (at least, that's the behavior I believe we had early on in v2 SoTW and it might be expected reasonably in a system that relies on some amount of eventual consistency.

Reply:
That sounds good - certainly my code before this PR agrees with that approach.

Let me just make sure I understand the sort of case you're describing.

Both before and after this PR, a (e.g.) CLA that none of our watches cares about is ignored, rather than rejecting the update. What this PR now rejects is the case where we have no EDS watches at all, and the server sent that CLA.

So, are you ok with treating "we have no watches at all" as a not-at-all-special case of "we have no watches interested in that resource"?

Reply:
Yeah, because I think if you consider the edge case of a single cluster, we can still race around the delivery of the first CLA, so I would not do the reject in that case.

Reply:
Great, thanks for confirming, done.